Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h

```diff
@@ -0,0 +1,1886 @@
+typename boost::graph_traits<PolygonMesh>::halfedge_descriptor
+random_halfedge_in_mesh(const PolygonMesh& pm, CGAL::Random& rnd = get_default_random())
+{
+  return internal::random_entity_in_range(halfedges(pm), rnd);
+}
+
+/// \brief returns a random non-null edge of the polygon mesh `pm`.
+///
+/// \tparam PolygonMesh A model of `EdgeListGraph`
+///
+template<typename PolygonMesh>
+typename boost::graph_traits<PolygonMesh>::edge_descriptor
+random_edge_in_mesh(const PolygonMesh& pm, CGAL::Random& rnd = get_default_random())
+{
+  return internal::random_entity_in_range(edges(pm), rnd);
+}
+
+/// \brief returns a random non-null face of the polygon mesh `pm`.
+///
+/// \tparam PolygonMesh A model of `FaceListGraph`
+///
+template<typename PolygonMesh>
+typename boost::graph_traits<PolygonMesh>::face_descriptor
+random_face_in_mesh(const PolygonMesh& pm, CGAL::Random& rnd = get_default_random())
+{
+  return internal::random_entity_in_range(faces(pm), rnd);
+}
+
+/// \brief returns the number of `next` one has to apply to the halfedge `hd`
+///        for `source(hd, mesh) == vd` to be true, starting from `hd = halfedge(fd, tm)`.
+///
+/// \tparam PolygonMesh A model of `FaceGraph`
+///
+/// \param vd a vertex of `pm` whose index we seek
+/// \param fd a face of `pm` in which we seek the index of `vd`
+/// \param pm a mesh of type `PolygonMesh`
+///
+/// \pre `vd` is a vertex of `fd`.
+template <typename PolygonMesh>
+int vertex_index_in_face(const typename boost::graph_traits<PolygonMesh>::vertex_descriptor vd,
+                         const typename boost::graph_traits<PolygonMesh>::face_descriptor fd,
+                         const PolygonMesh& pm)
+{
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_descriptor halfedge_descriptor;
+
+  halfedge_descriptor start = halfedge(fd, pm);
+  halfedge_descriptor current = start;
+  int counter = 0;
+
+  do
+  {
+    if(source(current, pm) == vd)
+      break;
+
+    ++counter;
+    current = next(current, pm);
+  }
+  while(current != start);
+
+  CGAL_postcondition(counter == 0 || current != start);
+
+  return counter;
+}
+
+/// \brief returns the number of `next` one has to apply to `hd` for `hd == he`
+///        to be true, starting from `hd = halfedge(face(hd, tm), tm)`.
+///
+/// \tparam PolygonMesh A model of `FaceGraph`.
+///
+/// \param he a halfedge of `pm` whose index in `face(he, tm)` we seek
+/// \param pm an object of type `PolygonMesh`
+///
+template <typename PolygonMesh>
+int halfedge_index_in_face(typename boost::graph_traits<PolygonMesh>::halfedge_descriptor he,
+                           const PolygonMesh& pm)
+{
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_descriptor halfedge_descriptor;
+  typedef typename boost::graph_traits<PolygonMesh>::face_descriptor     face_descriptor;
+
+  CGAL_precondition(he != boost::graph_traits<PolygonMesh>::null_halfedge());
+  CGAL_precondition(!is_border(he, pm));
+
+  face_descriptor f = face(he, pm);
```

Comment:
check that the faces are the same? + special case for border?

Reply:
Which two faces? There is only one face, described by the halfedge passed in parameter.

Reply:
My bad, I misread.